python/simulation.py

```diff
@@ -4425,6 +4425,41 @@ def _stop(sim):
 
     return _stop
 
+def stop_when_dft_decayed(decay_by, minimum_run_time=0, maximum_run_time=None):
+    """
+    Return a `condition` function, suitable for passing to `sim.run()` as the `until`
+    or `until_after_sources` parameter, that checks all of the Simulation's dft objects
+    every `dt` timesteps, and stops the simulation once all the components and frequencies
+    of *every* dft object have decayed by at least `decay_by`. The routine calculates the optimal
+    `dt` to use based on the frequency content in the DFT monitors.
+
+    Optionally the user can specify a minimum run time (`minimum_run_time`) or a maximum
+    run time (`maximum_run_time`).
+    """
+    # Record data in closure so that we can persitently edit
+    closure = {'previous_fields':1, 't0':0, 'dt':0}
+    def _stop(_sim):
+        if _sim.fields.t == 0:
+            closure['dt'] = max(int(1/_sim.fields.dft_maxfreq()/_sim.fields.dt),int(_sim.fields.min_decimation()))
+        if _sim.fields.t <= closure['dt'] + closure['t0']:
+            return False
+        elif maximum_run_time and _sim.round_time() > maximum_run_time:
+            return True
+        else:
+            previous_fields = closure['previous_fields']
+            current_fields  = _sim.fields.dft_norm2()
+            relative_change = np.abs(previous_fields-current_fields) /  previous_fields   
+            
+            closure['previous_fields'] = current_fields
+            closure['t0'] = _sim.fields.t
+
+            if mp.verbosity > 0:
+                fmt = "DFT decay(t = {0:1.1f}): {1:0.4e}"
+                print(fmt.format(_sim.meep_time(), np.real(relative_change)))
+            return relative_change <= decay_by and _sim.round_time(
```

Comment:
I'm not completely happy with the fact that the `decay_by` tolerance implicitly depends on `closure['dt']`.

(In contrast, my alternate proposal of `|fields| ≤ tol * max|fields|` has a `tol` that is independent of the sampling frequency.)

One alternative would be to check `change <= tol * maxchange`, where `change = np.abs(previous_fields-current_fields)` and `maxchange` is the maximum observed change, so that `tol` becomes independent of the sampling frequency.

Reply:
Why does `decay_by` implicitly depend on `closure['dt']`? We are only looking at a predetermined set of frequencies. So long as the frequencies are below the Nyquist limit, the sample rate shouldn't affect the decay tolerance, right?

> One alternative would be to check change <= tol * maxchange, where change = np.abs(previous_fields-current_fields) and maxchange is the maximum observed change over the simulation, so that tol becomes independent of the sampling frequency.

Aren't we already normalizing out any dependence? How different is the current implementation:
```relative_change = np.abs(previous_fields-current_fields) /  previous_fields ```

from your proposal:
```change = np.abs(previous_fields-current_fields) /  maxchange ```

The issue with `maxchange` is now you have to compute a Δ over every spatial point, which requires *caching* all of the dft fields.

Reply:
No, by `maxchange` I mean the maximum value of `change` observed over all the previous sampling *times*.

Reply:
Fixed.